cobalt/persistent_storage/persistent_settings.cc

```diff
@@ -70,171 +71,116 @@ void PersistentSettings::InitializePrefStore() {
-  base::task_runner_util::WaitForFence(thread_.task_runner(), FROM_HERE);
-  base::AutoLock auto_lock(pref_store_lock_);
-  base::Value::Dict* result = persistent_settings_.FindDict(key);
-  base::flat_map<std::string, std::unique_ptr<base::Value>> dict;
-  if (result) {
-    for (base::Value::Dict::iterator it = result->begin(); it != result->end();
-         ++it) {
-      dict.insert(std::make_pair(
-          it->first, base::Value::ToUniquePtrValue(std::move(it->second))));
-    }
-    return dict;
+  pref_store_->SetValue(key, value.Clone(),
+                        WriteablePrefStore::LOSSY_PREF_WRITE_FLAG);
+  if (validated_initial_settings_) {
+    CommitPendingWrite();
   }
-  return dict;
 }
 
-void PersistentSettings::SetPersistentSetting(
-    const std::string& key, std::unique_ptr<base::Value> value, bool blocking) {
-  task_runner()->PostTask(
-      FROM_HERE,
-      base::BindOnce(&PersistentSettings::SetPersistentSettingHelper,
-                     base::Unretained(this), key, std::move(value), blocking));
+void PersistentSettings::Remove(const std::string& key) {
+  if (!pref_store_) return;
+  if (base::SequencedTaskRunner::GetCurrentDefault() != task_runner()) {
+    Run(FROM_HERE, base::BindOnce(&PersistentSettings::Remove,
+                                  base::Unretained(this), key));
+    return;
+  }
+  pref_store_->RemoveValue(key, WriteablePrefStore::LOSSY_PREF_WRITE_FLAG);
+  if (validated_initial_settings_) {
+    CommitPendingWrite();
+  }
 }
 
-void PersistentSettings::SetPersistentSettingHelper(
-    const std::string& key, std::unique_ptr<base::Value> value, bool blocking) {
-  DCHECK_EQ(base::SequencedTaskRunner::GetCurrentDefault(), task_runner());
-  base::AutoLock auto_lock(pref_store_lock_);
-  pref_store_->SetValue(key, base::Value::FromUniquePtrValue(std::move(value)),
-                        WriteablePrefStore::DEFAULT_PREF_WRITE_FLAGS);
-  persistent_settings_ = pref_store_->GetValues();
+void PersistentSettings::RemoveAll() {
+  if (!pref_store_) return;
+  if (base::SequencedTaskRunner::GetCurrentDefault() != task_runner()) {
+    Run(FROM_HERE,
+        base::BindOnce(&PersistentSettings::RemoveAll, base::Unretained(this)));
+    return;
+  }
+  base::Value::Dict settings = pref_store_->GetValues();
+  for (auto it : settings) {
+    pref_store_->RemoveValue(it.first,
+                             WriteablePrefStore::LOSSY_PREF_WRITE_FLAG);
+  }
   if (validated_initial_settings_) {
-    CommitPendingWrite(blocking);
+    CommitPendingWrite();
   }
 }
 
-void PersistentSettings::RemovePersistentSetting(const std::string& key,
-                                                 bool blocking) {
-  task_runner()->PostTask(
-      FROM_HERE,
-      base::BindOnce(&PersistentSettings::RemovePersistentSettingHelper,
-                     base::Unretained(this), key, blocking));
+base::Value::Dict PersistentSettings::dict() const {
+  return pref_store_ ? pref_store_->GetValues() : base::Value::Dict();
```

Comment:
`dict()` gets called from different threads, but `GetValues()` is not thread-safe and `pref_store_` gets mutated without synchronization with this access.